src/volume_rendering/volume_render_layer.ts

```diff
@@ -242,31 +242,59 @@ export class VolumeRenderingRenderLayer extends PerspectiveViewRenderLayer {
 `;
           let glsl_emitIntensity = `
 void emitIntensity(float value) {
+}
+void setDefaultIntensityIfNotSet(float value) {
 }`;
+
           let glsl_continualEmit = ``;
-          if (isProjection(shaderParametersState.mode)) {
+          if (isProjectionMode(shaderParametersState.mode)) {
             builder.addFragmentCode(`
-float newIntensity = 0.0;
+float userIntensity = -1.0;
+float defaultIntensity = -1.0;
 float savedDepth = 0.0;
 float savedIntensity = 0.0;
 `);
-            glsl_emitIntensity = `
+            const emitFunctions = `
 void emitIntensity(float value) {
-  newIntensity = clamp(value, 0.0, 1.0);
-}`;
-            if (shaderParametersState.mode === VOLUME_RENDERING_MODES.MIN) {
+  userIntensity = convertIntensity(value);
+}
+void setDefaultIntensityIfNotSet(float value) {
+  float intensitySet = step(-0.00001, defaultIntensity);
+  defaultIntensity = mix(convertIntensity(value), defaultIntensity, intensitySet);
+}
+float getIntensity() {
+  float userIntensitySet = step(-0.00001, userIntensity);
+  return mix(defaultIntensity, userIntensity, userIntensitySet);
+}
+void resetIntensity() {
+  userIntensity = -1.0;
+  defaultIntensity = -1.0;
+}
+`;
+            glsl_emitIntensity = `
+float convertIntensity(float value) {
+  return clamp(value, 0.0, 1.0);
+}
+${emitFunctions}
+`;
```

Comment:
This is to handle a default intensity which would be defaulted to in min/max mode, which can be overwritten if the user calls `emitIntensity`.

Reply:
Can you comment on the reason for mixing and these various epsilon values?

Reply:
I've now removed most of these checks with the new heuristic we are using, where the value of the invlerp is the default intensity. Previously, the checks were to handle this.

One check remains, which I think is needed and can't be avoided. That is to compare the intensity value from the current ray position against the max intensity encountered along a ray in this chunk and pick the strictly bigger one. Then, if the intensity changed (i.e the new intensity was bigger), update the saved Color and Depth values.

There are two ways I can think of to implement this, and if you are aware of something clearer or more efficient @jbms that would be great!

Option 1: Step and Mix pattern (see [this branch](https://github.com/MetaCell/neuroglancer/tree/feature/max-with-step-and-mix))

```glsl
float intensityChanged = step(savedIntensity, newIntensity - 0.00001);
savedIntensity = mix(savedIntensity, newIntensity, intensityChanged); 
savedDepth = mix(savedDepth, depthAtRayPosition, intensityChanged);
outputColor = mix(outputColor, newColor, intensityChanged);
```

Option 2: Ternary operator (on this branch):

```glsl
bool intensityChanged = newIntensity > savedIntensity;
savedIntensity = intensityChanged ? newIntensity : savedIntensity; 
savedDepth = intensityChanged ? depthAtRayPosition : savedDepth;
outputColor = intensityChanged ? newColor : outputColor;
```

If you have a preference based on performance/code clarity etc. or have a suggestion just let me know.

Thanks!